### PR TITLE
Water - HOST Tensor AVX2 Support and Vectorized HIP support 

### PR DESCRIPTION
**Pull request by sampath1117**
*added for U8 , F32 , I8 , F16 variants
*added for PKD3,PLN3,PLN1 with toggle
*added test case for water in new test suite
*added golden outputs for water

### CODE REVIEW

**Comment by sampath1117** [comment]
@r-abishek 
For water the QA tests are failing currently. I need to debug more on this

Please note

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 1]
```diff
@@ -3101,16 +3101,21 @@ inline void rpp_resize_nn_load_i8pln1(Rpp8s *srcRowPtrsForInterp, Rpp32s *loc, _
 
 inline void rpp_store12_u8pkd3_to_u8pln3(Rpp8u* dstPtrR, Rpp8u* dstPtrG, Rpp8u* dstPtrB, __m128i &p)
 {
-    _mm_storeu_si128((__m128i *)(dstPtrR), _mm_shuffle_epi8(p, xmm_char_maskR)); /* Shuffle and extract the R pixels*/
-    _mm_storeu_si128((__m128i *)(dstPtrG), _mm_shuffle_epi8(p, xmm_char_maskG)); /* Shuffle and extract the G pixels*/
-    _mm_storeu_si128((__m128i *)(dstPtrB), _mm_shuffle_epi8(p, xmm_char_maskB)); /* Shuffle and extract the B pixels*/
+    _mm_storeu_si32((__m128i *)(dstPtrR), _mm_shuffle_epi8(p, xmm_char_maskR)); /* Shuffle and extract the R pixels*/
```
[the remark this answers is not in the document]
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_simd.hpp:3104]: Yes

File: src/modules/cpu/kernel/water.hpp [hunk 2]
```diff
@@ -0,0 +1,968 @@
...
+                                  RpptDescPtr dstDescPtr,
+                                  Rpp32f *amplitudeXTensor,
+                                  Rpp32f *amplitudeYTensor,
+                                  Rpp32f *frequencyXTensor,
+                                  Rpp32f *frequencyYTensor,
+                                  Rpp32f *phaseXTensor,
+                                  Rpp32f *phaseYTensor,
+                                  RpptROIPtr roiTensorPtrSrc,
+                                  RpptRoiType roiType,
+                                  RppLayoutParams layoutParams)
```
[the remark this answers is not in the document]
Comment (sampath1117) [reply src/modules/cpu/kernel/water.hpp]: Done

File: src/modules/cpu/kernel/water.hpp [hunk 3]
```diff
@@ -0,0 +1,968 @@
...
+        Rpp32u alignedLength = bufferLength & ~3;   // Align dst width to process 4 dst pixels per iteration
+        Rpp32s srcLocArray[4] = {0};         // Since 4 dst pixels are processed per iteration
+        Rpp32s invalidLoad[4] = {0};         // Since 4 dst pixels are processed per iteration
+
+        __m128 pSrcStrideH = _mm_set1_ps(srcDescPtr->strides.hStride);
+        __m128 pRoiLTRB[4];
+        pRoiLTRB[0] = _mm_set1_ps(roiLTRB.ltrbROI.lt.x);
+        pRoiLTRB[1] = _mm_set1_ps(roiLTRB.ltrbROI.lt.y);
+        pRoiLTRB[2] = _mm_set1_ps(roiLTRB.ltrbROI.rb.x);
+        pRoiLTRB[3] = _mm_set1_ps(roiLTRB.ltrbROI.rb.y);
```
[the remark this answers is not in the document]
Comment (sampath1117) [reply src/modules/cpu/kernel/water.hpp]: Changed all code to AVX2

**Comment by r-abishek** [comment]
@sampath1117 Is this ready now?

**Comment by sampath1117** [comment]
Hi @r-abishek 
I have made all the changes and the PR is ready for internal review

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 4]
```diff
@@ -1234,19 +1238,30 @@ inline void rpp_load24_f32pln3_to_f32pln3_mirror_avx(Rpp32f *srcPtrR, Rpp32f *sr
...
+    pRow[1] = _mm256_shuffle_ps(pTemp[0], pTemp[1], 0xDD);
+    pRow[2] = _mm256_shuffle_ps(pTemp[2], pTemp[3], 0x88);
+    pRow[3] = _mm256_shuffle_ps(pTemp[2], pTemp[3], 0xDD);
+
     __m128 p128[4];
-    p128[0] = _mm256_extractf128_ps(p[0], 0);
-    p128[1] = _mm256_extractf128_ps(p[1], 0);
-    p128[2] = _mm256_extractf128_ps(p[2], 0);
-    _MM_TRANSPOSE4_PS(p128[0], p128[1], p128[2], p128[3]);
+    p128[0] = _mm256_castps256_ps128(pRow[0]);
```
Comment (r-abishek) [inline src/include/cpu/rpp_cpu_simd.hpp:1252]: Please add some inline comments for better readability on vectorized code

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 5]
```diff
@@ -3101,16 +3231,78 @@ inline void rpp_resize_nn_load_i8pln1(Rpp8s *srcRowPtrsForInterp, Rpp32s *loc, _
...
+    _mm_storeu_si128((__m128i *)(dstPtrG + 5), _mm_shuffle_epi8(p128[1], maskG2));
+    _mm_storeu_si128((__m128i *)(dstPtrB + 5), _mm_shuffle_epi8(p128[1], maskB2));
+}
+
+inline void rpp_store24_i8pkd3_to_i8pln3_avx(Rpp8s* dstPtrR, Rpp8s* dstPtrG, Rpp8s* dstPtrB, __m256i &p)
+{
+    __m128i p128[2];
+    const __m128i maskR1 = _mm_setr_epi8(0, 3, 6, 9, 12, 15, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80);
+    const __m128i maskG1 = _mm_setr_epi8(1, 4, 7, 10, 13, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80);
+    const __m128i maskB1 = _mm_setr_epi8(2, 5, 8, 11, 14, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80);
```
Comment (r-abishek) [inline src/include/cpu/rpp_cpu_simd.hpp:3264]: Same comment here for reader to understand masks - say as to why maskR1 has  6 Rs and maskR2 has 2 Rs etc

File: src/modules/cpu/kernel/water.hpp [hunk 6]
```diff
@@ -0,0 +1,977 @@
...
+            {
+                Rpp32f *dstPtrTempR, *dstPtrTempG, *dstPtrTempB;
+                dstPtrTempR = dstPtrRowR;
+                dstPtrTempG = dstPtrRowG;
+                dstPtrTempB = dstPtrRowB;
+
+                Rpp32f dstX, dstY, sinFactor;
+                __m256 pDstX, pDstY, pSinFactor;
+                dstY = (Rpp32f)i;
+                sinFactor= std::sin((freqX * dstY) + phaseX);
```
Comment (r-abishek) [inline src/modules/cpu/kernel/water.hpp:366]: space before =

**Comment by r-abishek** [comment]
Looks okay, but again, do we have CSVs here for passing QA tests with/without tolerance?

File: src/include/cpu/rpp_cpu_simd.hpp (hunk 4 above)
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_simd.hpp]: Done

File: src/modules/cpu/kernel/water.hpp (hunk 6 above)
Comment (sampath1117) [reply src/modules/cpu/kernel/water.hpp]: Done

File: src/include/cpu/rpp_cpu_simd.hpp (hunk 5 above)
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_simd.hpp]: Done

**Review by r-abishek — APPROVED** [review]
lgtm

**Pull request merged**